execgen: remove COPYSLICE and WINDOW directives in favor of methods

yuzefovich opened the pull request:
**coldata: ensure Len() methods are inlined**

Release note: None

**execgen: remove COPYSLICE directive in favor of the method**

This commit removes `execgen.COPYSLICE` directive in favor of
newly-introduced `CopySlice` method on the well-typed vectors. Most of
these methods are inlined (decimals are an exception in some cases).

Additionally, this commit removes some leftover stuff.

Release note: None

**execgen: remove WINDOW directive in favor of the method**

All calls of the newly-introduced method are actually inlined.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66572)
<!-- Reviewable:end -->

yuzefovich committed:
coldata: ensure Len() methods are inlined

Release note: None

yuzefovich committed:
execgen: remove COPYSLICE directive in favor of the method

This commit removes `execgen.COPYSLICE` directive in favor of
newly-introduced `CopySlice` method on the well-typed vectors. Most of
these methods are inlined (decimals are an exception in some cases).

Additionally, this commit removes some leftover stuff.

Release note: None

yuzefovich committed:
execgen: remove WINDOW directive in favor of the method

All calls of the newly-introduced method are actually inlined.

Release note: None

yuzefovich commented:
Benchmarks of Copy:
```
name                                             old time/op    new time/op     delta
Copy/bytes/CopySimple/NullProbability=0.0-24        229ns ± 0%      228ns ± 1%   -0.31%  (p=0.015 n=8+10)
Copy/bytes/CopyWithSel/NullProbability=0.0-24      17.2µs ± 0%     17.4µs ± 1%   +1.13%  (p=0.037 n=9+10)
Copy/bytes/CopySimple/NullProbability=0.2-24        295ns ± 0%      304ns ± 1%   +3.12%  (p=0.000 n=7+10)
Copy/bytes/CopyWithSel/NullProbability=0.2-24      18.5µs ± 0%     18.3µs ± 0%   -1.01%  (p=0.000 n=10+9)
Copy/decimal/CopySimple/NullProbability=0.0-24     11.6µs ± 0%     11.5µs ± 0%   -0.87%  (p=0.000 n=10+8)
Copy/decimal/CopyWithSel/NullProbability=0.0-24    17.2µs ± 0%     17.4µs ± 1%   +1.04%  (p=0.000 n=9+10)
Copy/decimal/CopySimple/NullProbability=0.2-24     11.7µs ± 1%     11.5µs ± 0%   -2.35%  (p=0.000 n=10+10)
Copy/decimal/CopyWithSel/NullProbability=0.2-24    16.3µs ± 1%     16.9µs ± 1%   +3.56%  (p=0.000 n=9+9)
Copy/int/CopySimple/NullProbability=0.0-24          134ns ± 1%      133ns ± 0%   -0.60%  (p=0.003 n=10+10)
Copy/int/CopyWithSel/NullProbability=0.0-24         898ns ± 0%      801ns ± 0%  -10.79%  (p=0.000 n=10+8)
Copy/int/CopySimple/NullProbability=0.2-24          208ns ± 0%      202ns ± 0%   -2.55%  (p=0.000 n=10+10)
Copy/int/CopyWithSel/NullProbability=0.2-24        2.29µs ± 1%     2.40µs ± 0%   +4.95%  (p=0.000 n=10+10)

name                                             old speed      new speed       delta
Copy/bytes/CopySimple/NullProbability=0.0-24     35.7GB/s ± 0%   35.9GB/s ± 1%   +0.35%  (p=0.022 n=9+10)
Copy/bytes/CopyWithSel/NullProbability=0.0-24     476MB/s ± 0%    471MB/s ± 1%   -1.11%  (p=0.041 n=9+10)
Copy/bytes/CopySimple/NullProbability=0.2-24     27.8GB/s ± 0%   26.9GB/s ± 1%   -3.10%  (p=0.000 n=9+10)
Copy/bytes/CopyWithSel/NullProbability=0.2-24     442MB/s ± 0%    447MB/s ± 0%   +1.02%  (p=0.000 n=10+9)
Copy/decimal/CopySimple/NullProbability=0.0-24    704MB/s ± 0%    710MB/s ± 0%   +0.88%  (p=0.000 n=10+8)
Copy/decimal/CopyWithSel/NullProbability=0.0-24   477MB/s ± 0%    472MB/s ± 1%   -1.02%  (p=0.000 n=9+10)
Copy/decimal/CopySimple/NullProbability=0.2-24    697MB/s ± 1%    714MB/s ± 0%   +2.40%  (p=0.000 n=10+10)
Copy/decimal/CopyWithSel/NullProbability=0.2-24   503MB/s ± 1%    486MB/s ± 1%   -3.43%  (p=0.000 n=9+9)
Copy/int/CopySimple/NullProbability=0.0-24       61.2GB/s ± 1%   61.5GB/s ± 0%   +0.47%  (p=0.005 n=10+10)
Copy/int/CopyWithSel/NullProbability=0.0-24      9.13GB/s ± 0%  10.23GB/s ± 0%  +12.15%  (p=0.000 n=10+10)
Copy/int/CopySimple/NullProbability=0.2-24       39.4GB/s ± 0%   40.5GB/s ± 0%   +2.63%  (p=0.000 n=10+10)
Copy/int/CopyWithSel/NullProbability=0.2-24      3.58GB/s ± 1%   3.42GB/s ± 0%   -4.72%  (p=0.000 n=10+10)

name                                             old alloc/op   new alloc/op    delta
Copy/bytes/CopySimple/NullProbability=0.0-24        0.00B           0.00B          ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.0-24       0.00B           0.00B          ~     (all equal)
Copy/bytes/CopySimple/NullProbability=0.2-24        0.00B           0.00B          ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.2-24       0.00B           0.00B          ~     (all equal)
Copy/decimal/CopySimple/NullProbability=0.0-24      0.00B           0.00B          ~     (all equal)
Copy/decimal/CopyWithSel/NullProbability=0.0-24     0.00B           0.00B          ~     (all equal)
Copy/decimal/CopySimple/NullProbability=0.2-24      0.00B           0.00B          ~     (all equal)
Copy/decimal/CopyWithSel/NullProbability=0.2-24     0.00B           0.00B          ~     (all equal)
Copy/int/CopySimple/NullProbability=0.0-24          0.00B           0.00B          ~     (all equal)
Copy/int/CopyWithSel/NullProbability=0.0-24         0.00B           0.00B          ~     (all equal)
Copy/int/CopySimple/NullProbability=0.2-24          0.00B           0.00B          ~     (all equal)
Copy/int/CopyWithSel/NullProbability=0.2-24         0.00B           0.00B          ~     (all equal)

name                                             old allocs/op  new allocs/op   delta
Copy/bytes/CopySimple/NullProbability=0.0-24         0.00            0.00          ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.0-24        0.00            0.00          ~     (all equal)
Copy/bytes/CopySimple/NullProbability=0.2-24         0.00            0.00          ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.2-24        0.00            0.00          ~     (all equal)
Copy/decimal/CopySimple/NullProbability=0.0-24       0.00            0.00          ~     (all equal)
Copy/decimal/CopyWithSel/NullProbability=0.0-24      0.00            0.00          ~     (all equal)
Copy/decimal/CopySimple/NullProbability=0.2-24       0.00            0.00          ~     (all equal)
Copy/decimal/CopyWithSel/NullProbability=0.2-24      0.00            0.00          ~     (all equal)
Copy/int/CopySimple/NullProbability=0.0-24           0.00            0.00          ~     (all equal)
Copy/int/CopyWithSel/NullProbability=0.0-24          0.00            0.00          ~     (all equal)
Copy/int/CopySimple/NullProbability=0.2-24           0.00            0.00          ~     (all equal)
Copy/int/CopyWithSel/NullProbability=0.2-24          0.00            0.00          ~     (all equal)
```

jordanlewis reviewed (approved):
LGTM, this is great! I'll merge it since @yuzefovich is still OOO.

bors r+

mgartner commented:
Nice!





<!-- Sent from Reviewable.io -->
